File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        HashMap<Integer, HashSet<Integer>> adj = new HashMap();
        for (int i = 0; i < n; i++) adj.put(i, new HashSet());
        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        return dfs(adj, source, destination);
    }
    private boolean dfs(HashMap<Integer, HashSet<Integer>> adj, int s, int t) {
        if (s == t) return true;
        if (!adj.containsKey(s)) return false;
        HashSet<Integer> currentList = adj.get(s);
        adj.remove(s);
        for(var node : currentList) {
            if (dfs(adj, (int)node, t)) return true;
        }
        return false;
    }
 }

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.kt
class Solution {
    fun validPath(n: Int, edges: Array<IntArray>, source: Int, destination: Int): Boolean {
        val neighbors = Array<ArrayList<Int>>(n) { ArrayList<Int>() }
        val visited = Array<Boolean>(n) { false }
        for (i in edges.indices) {
            val a = edges[i][0]
            val b = edges[i][1]
            neighbors[a].add(b)
            neighbors[b].add(a)
        }
        return (dfs(neighbors, visited, source, destination))
           
    }
    fun dfs(neighbors: Array<ArrayList<Int>>, 
           visited: Array<Boolean>,
           start: Int,
           end: Int): Boolean {
        if (start == end) return true
        if (visited[start]) return false
        visited[start] = true
        for (e in neighbors[start]) {
            if (dfs(neighbors, visited, e, end)) return true
        }
        return false
    }
}